### PR TITLE
feat: extract budget intent via LLM

### PR DESCRIPTION
**Pull request by anandaroop**
Completes the budget inference step that was stubbed in #14044 

The LLM logic was developed over in https://github.com/artsy/quantum/pull/30 and is re-used in slightly modified form here.

Upon submitting messy free text such as “**idk maybe 5k by the end of the year**” the intent will be parsed as shown in the global state object here:

```js
{
  "currentStep": "result",
  "goal": "Decorate my home",

  // 👇🏽 messy input
  "budget": "idk maybe 5k by the end of the year",   

  // 👇🏽 extracted intent
  "budgetIntent": {                
    "budget": {
      "currency": "USD",
      "min": 0,
      "max": 5000
    },
    "timeFrame": {
      "number": 6,
      "unit": "month"
    }
  },

  "interests": [
    "Prints"
  ],
  "interestsFreeText": ""
}
```

https://github.com/artsy/force/assets/140521/bd9a9d5a-1644-4420-9983-64cedc343cbd

### CODE REVIEW

File: src/Apps/ArtAdvisor/07-Curated-Discovery/llm/extractBudgetIntent.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
+import { generateObject } from "ai"
+import { openai } from "@ai-sdk/openai"
+import dedent from "dedent"
+import { z } from "zod"
+
+const model = openai("gpt-4o")
```
Comment (anandaroop) [inline src/Apps/ArtAdvisor/07-Curated-Discovery/llm/extractBudgetIntent.ts:6]: The brand new Claude Sonnet 3.5 model appears to be [the new leader on many benchmarks](https://arstechnica.com/information-technology/2024/06/anthropics-latest-best-ai-model-is-twice-as-fast-and-still-terrible-at-dad-jokes/), but so far I observed 4o doing better on this task. Can maybe revisit.

File: src/Apps/ArtAdvisor/07-Curated-Discovery/App.tsx [hunk 2]
```diff
@@ -1,17 +1,16 @@
...
 import { Form } from "./Components/Form/Form"
 import { Result } from "./Components/Result/Result"
+import { schema as budgetIntentSchema } from "Apps/ArtAdvisor/07-Curated-Discovery/llm/extractBudgetIntent"
 
 type Step = "form" | "result"
 
-export type BudgetIntent = {
-  amount: number
-}
+export type BudgetIntent = z.infer<typeof budgetIntentSchema>
```
[the remark this answers is not in the document]
Comment (mc-jones) [reply src/Apps/ArtAdvisor/07-Curated-Discovery/App.tsx:9]: I like this, but I guess there is a slight trade off here in that the `schema` includes more information than just the typing, so devs will need to parse through that to understand the type.

File: src/Apps/ArtAdvisor/07-Curated-Discovery/Components/Result/Links.tsx [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+      >
+        Back
+      </Clickable>
+      <Clickable
+        textDecoration={"underline"}
+        py={1}
+        onClick={() => dispatch({ type: "RESET" })}
+      >
+        Start over
+      </Clickable>
```
[the remark this answers is not in the document]
Comment (mc-jones) [reply src/Apps/ArtAdvisor/07-Curated-Discovery/Components/Result/Links.tsx:25]: 🙏

File: src/Apps/ArtAdvisor/07-Curated-Discovery/llm/extractBudgetIntent.ts (hunk 1 above)
Comment (mc-jones) [reply src/Apps/ArtAdvisor/07-Curated-Discovery/llm/extractBudgetIntent.ts:6]: Nice. Will be interesting to test these more thoroughly for our use cases.

**Review by mc-jones — APPROVED** [review]
Few minor comments. But very cool. 🙏

File: src/Apps/ArtAdvisor/07-Curated-Discovery/App.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (anandaroop) [reply src/Apps/ArtAdvisor/07-Curated-Discovery/App.tsx:9]: That's a good point. I also thought it was a bummer that e.g. the Zod field-level description comments don't make it into the editor's intellisense via the generated TS types. But overall having a single SoT for the type information seemed to justify this compromise.

**Pull request merged**